### DISCUSSION TITLE
Running as a service

### DISCUSSION THREAD

**Discussion by k75Adventure** [nutdotnet/WinNUT-Client#161] [Q&A] [answered]
Based on the logs and simulated power outages, WinNut-Client isn't monitoring and initiating a shutdown while the computer is locked. 

Does anyone have any tips on how to run WinNUT-Client as a service?

I have tried using sc.exe create; however, I cannot get the service to run. I've had success using AlwaysUp, but I'd rather not rely on another application. 

Any suggestions would be greatly appreciated.

**Comment by gbakeman** [comment]
Hi there,

This is definitely a desired feature but not something that will be ready any time soon. We're currently tracking this in #46 which you can subscribe to keep updated. There are also some suggested workarounds there (nothing I can officially endorse), the latest one being using something called NSSM. I do want to mention that since I commented there, WinNUT no longer stores its settings in the Registry: rather, in a config file so I'm not sure how that will work now.

All that being said, I have WinNUT running and monitoring on a computer that's typically locked, so it sounds like there's something else going on here if you want to look into it. Could you tell me more about the issue?

**Reply by k75Adventure** [reply]
Hey @gbakeman, 

Thanks for the response! I did come across issue #46. I haven't tried NSSM yet; I've been trying to identify a solution that doesn't rely on a third-party tool. I decided to quickly test AlwaysUp to see if it would work.

I misspoke. WinNut continues to monitor when the user is locked. However, it doesn't monitor if the user isn't signed in at least based on the logs. I am trying to get WinNut-Client to shut down a Hyper-V Windows server. 

Thanks!

**Reply by gbakeman** [reply]
It sounds like you may have a similar use case to me - I have an old HP PC running Windows Server 2008 R2 which has a couple of Hyper-V VMs inside it. I have WinNUT running on the host, set to hibernate the computer if power goes down for long enough. Can you tell me why being locked but not signed in is causing issues for you?

Edit: Sorry, read your post a little more closely. If the user account logs out completely, then I'm afraid that would shut down WinNUT as all user programs are. I see your difficulty!

**Reply by k75Adventure** [reply]
Exactly! It sounds like something like NSSM or AlwaysUp is necessary to work around this current limitation of WinNut? 

I tried using Task Scheduler but wasn't successful.

**Reply by gbakeman** [reply]
I have to preface this by saying running a GUI program as a service is unknown territory to me, so take this with a grain of salt. WinNUT is a Windows Forms app first and foremost, so it depends on having that user interface available while it's running. Once you have it configured, you'll probably need an app to encapsulated WinNUT so it's able to display its GUI without running into any issues. I would say to go with NSSM since the other user mentioned it working fine, if AlwaysUp doesn't do it for you. And please let us know what you find out!

**Reply by k75Adventure** [reply]
Thanks so much; I really appreciate your input and suggestions!

**Reply by k75Adventure** [reply, marked as answer]
Just following up, NSSM worked perfectly. Created a service and ran it as a user.

**Reply by gbakeman** [reply]
Glad to hear it worked! Make sure you subscribed to the issue (#46) and you'll start getting pings once work is underway on making this a native feature.

**Reply by k75Adventure** [reply]
I definitely will; I look forward to that! Thanks for your work on the project. It's super useful and needed!
